Restrict `NameSegment` operations

sellout opened the pull request:
## Overview

With `OverloadedStrings` enabled globally and an `IsString` instance, the `newtype` was rendered useless.

## Implementation notes

This removes the `IsString` instance, extracts the `NameSegment` constructor/eliminator to a `.Internal` module and has `Unison.NameSegment` only re-export the type and sentinel segment names.

This forces cascading changes, including eliminating a bunch of magic literals scattered throughout the code.

sellout committed:
Restrict `NameSegment` operations

With `OverloadedStrings` enabled globally and an `IsString` instance, the
`newtype` was rendered useless.

This extracts the `NameSegment` constructor/eliminator to a `.Internal` module,
has `Unison.NameSegment` only re-export the type, and moves the `*Segment`
members to `Unison.Syntax.NameSegment`.

This forces cascading changes, including eliminating a bunch of magic literals
scattered throughout the code.

aryairani commented:
@mitchellwrosen Could you review this one too

mitchellwrosen commented on codebase2/core/Unison/NameSegment.hs:
```diff
@@ -1,57 +1,15 @@
-docSegment =
-  "doc"
-
+--  __TODO__: This should live in "Unison.Syntax.NameSegment", but it’s currently used in unison-core.
```
As discussed, we are going to pull docSegment and friends back in here

mitchellwrosen commented on unison-cli/src/Unison/Codebase/Editor/UriParser.hs:
```diff
@@ -62,9 +62,12 @@ type P = P.Parsec Void Text.Text
-    ReadRemoteNamespaceGit <$> readGitRemoteNamespace
-      <|> ReadShare'ProjectBranch <$> projectAndBranchNamesParserInTheContextOfAlsoParsingLooseCodePaths specifier
-      <|> ReadShare'LooseCode <$> readShareLooseCode
+    ReadRemoteNamespaceGit
```
ormolu seems to have gotten the precedence wrong here and below?

aryairani commented:
Will y'all signal when the review comments are addressed to your respective satisfaction?

mitchellwrosen commented:
@aryairani yeah – we reviewed over the phone, it is good to go aside from having conflicts and some ormolu things to work out

aryairani commented:
Ok — go ahead and merge then when ready

sellout committed:
Add a `TypeError` for `IsString NameSegment`

We don’t want someone to come along and add this instance without thinking about it.

sellout committed:
Move sentinel segment names to `Unison.NameSegment`

They don’t belong in `Unison.Syntax.NameSegment` because they are not
part of the syntax – they are names stored in the codebase that need to
match regardless of the syntax used.

sellout committed:
Fix formatting that implies wrong precedence

I think this was due to an Ormolu version mismatch. Ormolu is happy with
this change, so I think it was originally refomatted before I was using
Ormolu from the Unison Nix environment.

sellout committed:
Merge remote-tracking branch 'upstream/trunk' into restrict-NameSegment

sellout commented:
Ok, should be all good. And merged in trunk.

mitchellwrosen commented:
Looks like some tests have compiler errors @sellout

sellout commented:
> Looks like some tests have compiler errors @sellout

Sorry – during merge conflict resolution I even said to myself “I’m going to let the compiler help me get this one right“, then forgot to build the tests before pushing. It should be fixed now (since I actually built & ran the tests).
